parsl/providers/azure/azure.py

```diff
@@ -0,0 +1,489 @@
+import base64
+
+from parsl.dataflow.error import ConfigurationError
+from parsl.providers.azure.template import template_string
+from parsl.providers.provider_base import ExecutionProvider
+from parsl.providers.error import OptionalModuleMissing
+from parsl.utils import RepresentationMixin
+from parsl.launchers import SingleNodeLauncher
+
+logger = logging.getLogger(__name__)
+
+try:
+    from azure.common.credentials import ServicePrincipalCredentials
+    from azure.mgmt.resource import ResourceManagementClient
+    from azure.mgmt.network import NetworkManagementClient
+    from azure.mgmt.compute import ComputeManagementClient
+    from azure.mgmt.compute.models import DiskCreateOption
+
+    _api_enabled = True
+
+except ImportError:
+    _api_enabled = False
+else:
+    _api_enabled = True
+
+translate_table = {
+    'VM pending': 'PENDING',
+    'VM running': 'RUNNING',
+    'VM deallocated': 'COMPLETED',
+    'VM stopping': 'COMPLETED',  # We shouldn't really see this state
+    'VM stopped': 'COMPLETED',  # We shouldn't really see this state
+}
+
+
+class AzureProvider(ExecutionProvider, RepresentationMixin):
+    """
+    A Provider for using Microsoft Azure Resources
+
+    One of 2 methods are required to authenticate: keyfile, or environment
+    variables. If  keyfile is not set, the following environment
+    variables must be set: `AZURE_CLIENT_ID` (the access key for
+    your azure account),
+    `AZURE_CLIENT_SECRET` (the secret key for your azure account), the
+    `AZURE_TENANT_ID` (the session key for your azure account), and
+    AZURE_SUBSCRIPTION_ID.
+
+    Parameters
+    ----------
+    vm_reference : dict
+        Dictionary describing the parameters of the Azure VM.
+
+        Required structure:
+        {
+          'publisher': VM OS publisher
+          'offer': VM OS offer
+          'sku': VM OS SKU
+          'version': VM OS version
+          'vm_size': VM Size, analogous to instance type for AWS
+          'disk_size_gb': (int) size of VM disk in gb
+          "admin_username": (str) admin username of VM instances,
+          "password": (str) admin password for VM instances
+        }
+    worker_init : str
+        String to append to the Userdata script executed in the cloudinit phase of
+        instance initialization.
+    walltime : str
+        Walltime requested per block in HH:MM:SS.
+    key_file : str
+        Path to json file that contains 'Azure keys'
+    nodes_per_block : int
+        This is always 1 for Azure. Nodes to provision per block.
+    nodes_per_block : int
+        Nodes to provision per block. Default is 1.
+    init_blocks : int
+        Number of blocks to provision at the start of the run. Default is 1.
+    min_blocks : int
+        Minimum number of blocks to maintain. Default is 0.
+    max_blocks : int
+        Maximum number of blocks to maintain. Default is 10.
+    region : str
+        Azure region to launch machines. Default is 'westus'.
+    key_name : str
```

Comment:
does this get used anywhere?

Reply:
Azure requires an "admin password" to be set for Linux VM instances (which I really don't like) but optionally allows for SSH keys to be used as well. I am open to opinions on whether it would be useful to have keys implemented, knowing that passwords are required anyways.

Reply:
I'm a big fan of SSH keys, so I'm not going to say no.